### PR TITLE
fix minillm trainer

### PR DESCRIPTION
**Pull request by t1101675**
# What does this PR do?

Fix the convergence issues of MiniLLMTrainer. It can now smoothly optimize the Reverse KL Divergence between the teacher and the student:

image

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by t1101675** [comment]
@qgallouedec I have reduced #4731 to only fixing the convergence of the MiniLLM trainer.

File: trl/trainer/grpo_trainer.py [hunk 1]
```diff
@@ -1890,7 +1878,7 @@ def _generate_and_score_completions(
             generate_every = self.args.steps_per_generation * self.num_iterations  # generation frequency
             if self.args.gradient_accumulation_steps % generate_every != 0 or (
                 self.use_vllm and self.vllm_importance_sampling_correction
-            ):
+            ) or self.args.always_track_old_logps:
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py]: I'm not sure to understand why you need this? When `self.args.gradient_accumulation_steps % generate_every == 0`, then `old_per_token_logps == per_token_logps`, why not just using `per_token_logps.detach()`?

Comment (t1101675) [reply trl/trainer/grpo_trainer.py]: Nice comments! I have removed `always_track_old_logps`.

**Comment by donglixp** [comment]
@qgallouedec Any update on this PR?

**Comment by gathierry** [comment]
Hi @t1101675 @donglixp , I'm working on some experiments with minillm trainer. May I ask how this PR is going to fix the convergence? It seems like most updates are logging and profiling.
Is `on_policy_logq` the critical change?
And is it possible to share the configs of your experiment `minillm/openr1-math-220k/qwen2.5-1.5B/lr5e-6_Vllm_18192.`

My understanding is you set `on_policy_logq=False` then use `student_log_probs_on_labels = inputs["old_per_token_logps"]` to fix the convergence issue?
Please correct me if I'm wrong.

Thanks

**Comment by t1101675** [comment]
How this PR fixes convergence:
1. It corrects the way to compute the masks for teacher logprobs (this [line](https://github.com/huggingface/trl/pull/4743/changes#diff-c6698b0b9f7db58911abe8e3dc983cf2eb9a0ad48cf6aa8e092ec0ea6f090bfcR391))
2. It considers the gradient accumulation steps when computing the single_step_decomposition_loss (this [line](https://github.com/huggingface/trl/pull/4743/changes#diff-c6698b0b9f7db58911abe8e3dc983cf2eb9a0ad48cf6aa8e092ec0ea6f090bfcR434))

The config of `minillm/openr1-math-220k/qwen2.5-1.5B/lr5e-6_Vllm_l8192`:

```
#!/bin/bash


SLURM_NNODES=1
head_node_ip=localhost

export OMP_NUM_THREADS=32
export PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True
export VLLM_LOGGING_LEVEL=ERROR

if [[ -z $CUDA_VISIBLE_DEVICES ]]; then
    gpu_count=$(nvidia-smi --query-gpu=name --format=csv,noheader | wc -l)
else
    gpu_count=$(echo $CUDA_VISIBLE_DEVICES | tr ',' '\n' | wc -l)
fi
echo "Available GPU count: $gpu_count"

GROUP_NAME=minillm/openr1-math-220k/qwen2.5-1.5B/lr5e-6_vllm_l8192
JOB_TYPE=train
RUN_NAME=${JOB_TYPE}/${GROUP_NAME}
OUTPUT_DIR=results/${RUN_NAME}

NUM_PROCESS=$((gpu_count * SLURM_NNODES))

BATCH_SIZE=256
MICRO_BATCH_SIZE=2
GRAD_ACC=$((BATCH_SIZE / MICRO_BATCH_SIZE / NUM_PROCESS))

read -r -d '' cmd <<EOF
accelerate launch --config_file=trl/accelerate_configs/zero1.yaml --num_processes $NUM_PROCESS \
--num_machines $SLURM_NNODES --rdzv_backend c10d --main_process_ip $head_node_ip --main_process_port 29500 \
trl/scripts/minillm.py \
    --model_name_or_path results/train/sft/openr1-math-220k/qwen2.5-1.5B/lr4e-5/checkpoint-2695/ \
    --dtype bfloat16 \
    --teacher_model_name_or_path Qwen/Qwen2.5-3B-Instruct \
    --teacher_dtype bfloat16 \
    --dataset_name dataset/llm_rl/OpenR1-Math-220k \
    --learning_rate 5.0e-6 \
    --lr_scheduler_type cosine \
    --warmup_steps 100 \
    --num_train_epochs 1 \
    --per_device_train_batch_size $MICRO_BATCH_SIZE \
    --gradient_accumulation_steps $GRAD_ACC \
    --gradient_checkpointing \
    --eval_strategy no \
    --save_strategy steps \
    --save_steps 500 \
    --eval_steps 100 \
    --output_dir ${OUTPUT_DIR} \
    --attn_implementation flash_attention_2 \
    --resume_from_checkpoint True \
    --report_to none \
    --wandb_run_name $RUN_NAME \
    --wandb_mode online \
    --wandb_job_type $JOB_TYPE \
    --wandb_group $GROUP_NAME \
    --logging_steps 1 \
    --logging_first_step true \
    --dataset_num_proc 64 \
    --max_completion_length 8192 \
    --use_vllm \
    --vllm_mode colocate \
    --vllm_gpu_memory_utilization 0.05 \
    --rkl_advantage False
EOF

echo ${cmd}
bash -c "${cmd}"
```

**Comment by gathierry** [comment]
Thanks for the reply @t1101675 . I also notice this [update](https://github.com/huggingface/trl/pull/4743/changes#diff-c6698b0b9f7db58911abe8e3dc983cf2eb9a0ad48cf6aa8e092ec0ea6f090bfcR289) might be also important? Otherwise this [call](https://github.com/huggingface/trl/pull/4743/changes#diff-c6698b0b9f7db58911abe8e3dc983cf2eb9a0ad48cf6aa8e092ec0ea6f090bfcL345) will also be backwarded by RKL loss?

**Comment by t1101675** [comment]
Yes. Setting `torch.no_grad` for the advantage computation is also important.
